doc/oomd/SONiC OOM daemon.md

```diff
@@ -0,0 +1,255 @@
+# SONiC OOM daemon
+
+# Table of Contents
+- [Table of Contents](#table-of-contents)
+- [About this Manual](#about-this-manual)
+- [1 Functional Requirements](#1-functional-requirement)
+- [2 Configuration and Management Requirements](#2-configuration-and-management-requirements)
+  * [2.1 SONiC CLI](#21-sonic-cli)
+  * [2.2 Config DB](#22-config-db)
+- [3 Design](#design)
+  * [3.1 OOMD Implementation](#31-oomd-Implementation)
+  * [3.2 Default OOMD config](#32-default-oomd-config)
+  * [3.3 ConfigDB Schema](#34-configdb-schema)
+  * [3.4 CLI](#35-cli)
+- [4 Error handling](#error-handling)
+- [5 Serviceability and Debug](#serviceability-and-debug)
+- [6 Unit Test](#unit-test)
+- [8 References](#references)
+
+
+# About this Manual
+This document provides a detailed description on the new features for:
+ - Protect SONiC system memory by OOM daemon.
+ - OOM daemon config command.
+ - OOM daemon high level design.
+ - OOM daemon ConfigDB schema
+
+### SONiC memory issue sloved by this feature.
+ - Currently SONiC enabled OOM killer, and set /proc/sys/vm/panic_on_oom to 2, which will trigger kernal panic when OOM. This is by design to protect SONiC key process and container.
+ - A typical switch device have 4 GB memory and sonic usually will use 1.5 GB for dockers, and 500 MB for system process. so there will be 2 GB free memory for user. 
+ - sonic not enable swap for most device.
+ - When user run some command trigger OOM, SONiC will kernel panic. for example:
+   - Multiple user login to device, some service may create 10+ concurrent sessions login to device.
+   - Some user script/command take too much memory, currently 'show' command will take 70 MB memory.
+ - To fix this issue, we need 2 feature: login session limit and OOM daemon.
+
+# 1 Functional Requirement
+ - When exceed memory high-water mark, terminate user process to free memory and protect device from OOM happen.
+
+ - Can set system memory high-water mark and low-water mark.
+ - Can set memory protect policy:
+    - User/group list, their process can be terminate safely.
+    - Terminate policy: terminate user session or terminate user process.
+
+- All process start by local user will not be terminate by OOMD, unless they add to user/group list.
+- Provide default setting to minimize customer side change.
+  - Customers may have pipelines to initialize device configuration, because this feature add new commands, the pipeline may need update. The default limitation is designed to cover most scenario to minimize the customer side change.
```

Comment:
>pipelines [](http://example.com/codeflow?start=23&length=9)

Do you mean "network management software"?

Reply:
Fixed, change to network management software